1.17 fabric

Cjsah opened the pull request.

Cjsah committed:
21w20a

Cjsah committed:
fixed tooltip

Cjsah committed:
update

Cjsah committed:
complete

Cjsah committed:
complete

Cjsah committed:
Merge remote-tracking branch 'origin/1.17-fabric' into 1.17-fabric

Cjsah committed:
update

Cjsah committed:
update

Cjsah committed:
fix npe

SAGESSE-CN commented:
I don't think `modifyInt` and `OrderedTextTooltipComponentMixin` are necessary 

You can refer to this implementation:
https://github.com/squeek502/AppleSkin/blob/751321bc35d15f6480bd439e2ecf841206825827/java/squeek/appleskin/client/TooltipOverlayHandler.java#L74

Cjsah commented:
> I don't think `modifyInt` and `OrderedTextTooltipComponentMixin` are necessary
> 

I think `modifyInt` and `OrderedTextTooltipComponentMixin` are necessary

The purpose of `modifyInt` is to prevent this from happening 
![image](https://user-images.githubusercontent.com/46415647/120896537-9be59900-c654-11eb-90a1-efd13c276be5.png)

mojang uses `lines.stream().map(TooltipComponent::of).collect(Collectors.toList())` to convert to TooltipComponent.
I need `OrderedTextTooltipComponentMixin` to get FoodOverlay to get its length
Maybe I can solve this problem with Accessor to simplify it 

> You can refer to this implementation:
> https://github.com/squeek502/AppleSkin/blob/751321bc35d15f6480bd439e2ecf841206825827/java/squeek/appleskin/client/TooltipOverlayHandler.java#L74

Good idea! I will do it

Pardon my poor english

SAGESSE-CN commented:
This comment fixes the problem:
https://github.com/squeek502/AppleSkin/commit/69231c94f864711461a5a7c3508460f989457514#diff-0dc0e7f23822837654e553f62bc984eb9e909a69c40e3aaf1f8f76ee86333eae

Cjsah commented:
> This comment fixes the problem:
> [69231c9#diff-0dc0e7f23822837654e553f62bc984eb9e909a69c40e3aaf1f8f76ee86333eae](https://github.com/squeek502/AppleSkin/commit/69231c94f864711461a5a7c3508460f989457514#diff-0dc0e7f23822837654e553f62bc984eb9e909a69c40e3aaf1f8f76ee86333eae)

ok, I will delete these codes

Cjsah commented:
> This comment fixes the problem:
> [69231c9#diff-0dc0e7f23822837654e553f62bc984eb9e909a69c40e3aaf1f8f76ee86333eae](https://github.com/squeek502/AppleSkin/commit/69231c94f864711461a5a7c3508460f989457514#diff-0dc0e7f23822837654e553f62bc984eb9e909a69c40e3aaf1f8f76ee86333eae)

ok, I will delete `modifyInt`
but I think `OrderedTextTooltipComponentMixin` is necessary
Need to get `FoodOverlay` in the `onRenderTooltip` of the `TooltipOverlayHandler` 
I changed duck type to Accessor

Cjsah committed:
update

SAGESSE-CN commented:
> Need to get FoodOverlay in the onRenderTooltip of the TooltipOverlayHandler

You need to `OrderedTextTooltipComponentMixin` because return a new oredered text, this will increase the dependency of Minxin
```java
@Override		
public OrderedText asOrderedText() {
    return new FoodOverlayOrderedText(this.foodOverlay);
}
```

When return `this` in the `asOrderedText` method, the `List<? extends OrderedText> tooltip` by `onRenderTooltip` already included object kind of `FoodOverLayTextComponent` , not required additional convert.

Cjsah commented:
> > Need to get FoodOverlay in the onRenderTooltip of the TooltipOverlayHandler
> 
> You need to `OrderedTextTooltipComponentMixin` because return a new oredered text, this will increase the dependency of Minxin
> 
> ```java
> @Override		
> public OrderedText asOrderedText() {
>     return new FoodOverlayOrderedText(this.foodOverlay);
> }
> ```
> 
> When return `this` in the `asOrderedText` method, the `List<? extends OrderedText> tooltip` by `onRenderTooltip` already included object kind of `FoodOverLayTextComponent` , not required additional convert.

Yeah, I have changed it.
```java
@Mixin(OrderedTextTooltipComponent.class)
public interface OrderedTextTooltipComponentAccessor {
    @Accessor
    OrderedText getText();
}
```

And `FoodOverlayTextComponent` I have changed to this
https://github.com/squeek502/AppleSkin/blob/751321bc35d15f6480bd439e2ecf841206825827/java/squeek/appleskin/client/TooltipOverlayHandler.java#L74

SAGESSE-CN commented:
I mean `instanceof` is enough, there's no need to use `getText` 
```java
// Find food overlay of text lines.		
FoodOverlay foodOverlay = null;		
for (int i = 0; i < tooltip.size(); ++i) {
    if (tooltip.get(i) instanceof FoodOverlayTextComponent) {
        toolTipY += i * 10;				
        foodOverlay = ((FoodOverlayTextComponent)tooltip.get(i)).foodOverlay;
        break;
}
```

Although using `getText` is working, however `getText` needs to dependencies some not exported api, this is not a wise choice.

Cjsah commented:
> I mean `instanceof` is enough, there's no need to use `getText`
> 
> ```java
> // Find food overlay of text lines.		
> FoodOverlay foodOverlay = null;		
> for (int i = 0; i < tooltip.size(); ++i) {
>     if (tooltip.get(i) instanceof FoodOverlayTextComponent) {
>         toolTipY += i * 10;				
>         foodOverlay = ((FoodOverlayTextComponent)tooltip.get(i)).foodOverlay;
>         break;
> }
> ```

I don't think so
In `TooltipComponent`
```java
static TooltipComponent of(OrderedText text) {
        return new OrderedTextTooltipComponent(text);
    }
```
In `OrderedTextTooltipComponent `
```java
    private final OrderedText text;

    public OrderedTextTooltipComponent(OrderedText text) {
        this.text = text;
    }
```
> mojang uses `lines.stream().map(TooltipComponent::of).collect(Collectors.toList())` to convert to TooltipComponent.

It's a new class and `FoodOverlayTextComponent` is a member var of this class

SAGESSE-CN commented:
You are right, i'm sorry I got confused  :)

Cjsah commented:
You are still excellent, also thank you for helping me solve some problems

squeek502 reviewed (changes requested):
I'm not sure I can merge this as-is. When porting, please:

- Don't change the formatting of code, it just makes it harder to see what actually changed and messes up merging/rebasing/etc
- Don't make any changes that aren't absolutely necessary for the port. If there's something that could be improved, it should be done on the main branch and then merged into the newer branches afterwards
- Don't delete/change comments ([here's an example](https://github.com/squeek502/AppleSkin/pull/108/files#diff-308d92871c22a8dd347cc4c3bb55dca00700dd79bfecf42505cdac0d03d215d5L97))
- It would be good to have this rebased on the latest 1.16 branch (I realize that the branch has gotten a lot of changes recently which makes this harder)

squeek502 commented on java/squeek/appleskin/network/SyncHandler.java:
```diff
@@ -3,7 +3,7 @@
 import net.fabricmc.api.EnvType;
 import net.fabricmc.api.Environment;
-import net.fabricmc.fabric.api.network.ClientSidePacketRegistry;
+import net.fabricmc.fabric.impl.networking.ClientSidePacketRegistryImpl;
```
This looks like the wrong way to update this. The deprecation notice in `ClientSidePacketRegistry` says to migrate to `ClientPlayNetworking`. See also https://fabricmc.net/wiki/tutorial:networking

squeek502 commented:
Haven't figured it out fully yet, but I think the route we want to go with the tooltip stuff is to follow the example of `BundleTooltipComponent`:

- Have `FoodOverlayTextComponent` Implement `TooltipData` instead of `OrderedText` and add that to the tooltip
- Mixin `TooltipComponent.of(TooltipData)` to add something like `if (data instanceof AppleSkinTooltipData) return new AppleSkinTooltipComponent()`
- Implement the rendering in `AppleSkinTooltipComponent.drawItems`, allowing us to remove our `renderTooltip` mixins entirely (the entire `GuiMixin` class could be removed since we moved the `getTooltip` stuff to an `ItemStack` mixin in b2ed069e1f35c52b8f33dd341ef3c591ee773bfc)

EDIT: Nevermind, `TooltipData` is separate from the `List<Text>` and only allows for a single `Optional<TooltipData>`, so this plan won't work. Hopefully the TooltipData API gets fleshed out a bit more so we can use it in the future.

EDIT#2: Could do roughly the same thing but keep implementing `OrderedText` and mixin `TooltipComponent.of(OrderedText)` instead to return an `AppleSkinTooltipComponent`

EDIT#3: Mixin doesn't support static interface method injection: https://github.com/SpongePowered/Mixin/issues/318

squeek502 commented:
Thanks for this PR, but I ended up porting it while looking into the `TooltipData`/`TooltipComponent` stuff:

807c91b7a00bf7f63311dec53eb0d64cea992b69

Here's the comment describing how the solution I ended up going with works if you're interested:

https://github.com/squeek502/AppleSkin/blob/807c91b7a00bf7f63311dec53eb0d64cea992b69/java/squeek/appleskin/mixin/ScreenMixin.java#L22-L40

 I did not update the SyncHandler, though, so if you'd like to submit a PR that updates it to use `ClientPlayNetworking` I'd gladly accept it.
